.github/workflows/release-from-pr.yml

```diff
@@ -0,0 +1,303 @@
+          fetch-depth: 0
+          token: ${{ env.GITHUB_TOKEN }}
+      # TODO: Define if we use lock/unlock branch feature
+      #      - uses: elastic/apm-pipeline-library/.github/actions/branch-lock@feat/action-branch-lock
+      #        with:
+      #          lock: 'false'
+      #          branch: 'main'
+      #          token: ${{ secrets.PAT }}
+      - name: Merge the PR branch into the base branch
+        run: |
+          # Fetch the base branch
+          git fetch origin "${{ github.event.pull_request.base.ref }}"
+          git checkout "${{ github.event.pull_request.base.ref }}"
+          
+          # Try a fast-forward merge
+          # If it fails, the workflow will stop and the PR will not be merged
+          git merge --ff-only "${{ github.event.pull_request.head.ref }}"
+          
+          # Push the changes to the base branch
+          git push origin "${{ github.event.pull_request.base.ref }}"
+
+  prepare_release:
+    name: "Find tag & Wait propagation"
+    runs-on: ubuntu-latest
+    needs:
+      - check
+      - merge_ff
+    if: needs.check.outputs.is_release == 'true'
+    outputs:
+      tag_name: ${{ steps.tag.outputs.name }}
+    steps:
+      - id: tag
+        name: Find latest tag
+        uses: actions/github-script@v6
+        with:
+          result-encoding: string
+          retries: 3
+          script: |
+            const tags = await github.rest.repos.listTags({
+              owner: context.repo.owner,
+              repo: context.repo.repo,
+              per_page: 1
+            });
+            const tag = tags.data[0].name;
+            core.info(`Latest tag: ${tag}`);
+            core.setOutput('name', tag);
+
+      - name: Await artifacts published in maven central
+        shell: bash
+        timeout-minutes: 120
+        run: |
+          until .ci/release/wait_maven_artifact_published.sh "${TAG_NAME}"
+          do
+            echo "Artifacts not found on maven central. Sleeping 30 seconds, retrying afterwards"
+            sleep 30s
+          done
+
+  build_and_push_docker_images:
+    name: "Build and push docker images"
+    runs-on: ubuntu-latest
+    needs:
+      - check
+      - prepare_release
+    if: needs.check.outputs.is_release == 'true'
+    env:
+      SONATYPE_FALLBACK: 1
+    steps:
+      - name: Set tag name
+        run: |
+          echo "TAG_NAME=${{ needs.prepare_release.outputs.tag_name }}" >> ${GITHUB_ENV}
+      - uses: actions/checkout@v4
+        with:
+          ref: ${{ env.TAG_NAME }}
+          fetch-depth: 0 # Load entire history as it is required for the push-script
+      - uses: elastic/apm-pipeline-library/.github/actions/docker-login@current
+        with:
+          registry: docker.elastic.co
+          secret: secret/apm-team/ci/docker-registry/prod
+          url: ${{ secrets.VAULT_ADDR }}
+          roleId: ${{ secrets.VAULT_ROLE_ID }}
+          secretId: ${{ secrets.VAULT_SECRET_ID }}
+      - name: "Build docker image"
```

Comment:
I'd envision the `release-from-pr.yml` to consume the artifacts generated in the `release.yml`. This can help with ensuring that no artifacts need to be generated in two different places the ones in `release.yml` and the ones in `release-from-pr.yml`. But making the `release-from-pr.yml` to be the orchestrator.

Reply:
AFAIU, the docker build script will consume artifacts from the Sonatype repository since they aren't present locally.

Reply:
I'm not talking about the `docker build image` but a more generic aspect, I'd say the `release.yml` workflow shall be the one creating all the required artifacts, and the `release-from-pr.yml` workflow being the one promoting them - while `promotion` means interacting with Nexus, Docker, AWS or similar artifact managers.

Reply:
On side note regarding the `docker build script`, will it work when using the staging nexus repository? As I mentioned https://github.com/elastic/apm-agent-java/pull/3422#pullrequestreview-1727110547, this proposal will need to use the staging nexus repository, so releases won't be public available, only after the PR with the bumps is merged (that's the task in `release-from-pr.yml`). 

The promotion in Nexus from staging to production could be done in a follow up. The process for the Release Manager, aka RM, could be something like:

1) RM runs the `release.yml` workflow with the relevant input parameters
2) RM waits for some notification regarding the PR with the bumps to be approved.
  a) Once approved the `release-from-pr.yml` workflow will do the remaining steps.
3) RM promotes from stating to production in Nexus (manually if no automated).